### ISSUE TITLE
Enhancement: [@typescript-eslint/no-unnecessary-condition] Detect unnecessary comparison to falsy values

### ISSUE THREAD

**Issue by Samuel-Beslogic** [typescript-eslint/typescript-eslint#5021]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/

### Description

I would like unnecessary falsy values checks when the variable is already checked for truthiness in the same condition chain to be raised. I often see newer devs not realize that their condition is redundant.

### Fail

```typescript
const myVar = '' as string | number;

if (myVar && myVar !== '') {
  // do something
}

if (myVar && myVar !== 0) {
  // do something
}

if (myVar && !!myVar ) {
  // do something
}
```

### Pass

```typescript
const myVar = '' as string | number;

if (myVar) {
  // do something
}
```

**Comment by bradzacher** [comment]
Unfortunately this is not really achievable because the types do not convey this information.
You cannot *remove* specific cases from the primitive types:

```ts
declare const x: string;
// typeof x === string

if (!x) { return };
// typeof x === string

if (x !== '') { return }
// typeof x === string
```

So in order for this to be a lint rule, we would need to invent some method for tracking this by tracking a type as it progresses through a function's control-flow. This is obviously not really possible for us to build.

Something like negated types in TS (https://github.com/microsoft/TypeScript/pull/29317) would help solve this (because TS could refine the type to like `string & not ''`, but without it - we're SOL.

**Comment by Samuel-Beslogic** [comment]
Thank you for the explanation and reference. I went and upvoted it.
